### ISSUE TITLE
[Discuss] Referential integrity with Saved Objects

### ISSUE THREAD

**Issue by stacey-gammon** [stacey-gammon/kibana-dependency-tracking#43]
TLDR; This thread is to discuss extending Saved Objects service to add _referential integrity_.

---

Right now referential integrity in Kibana is not enforced. For example, you can delete a visualization, which is used by one or more dashboards, and those dashboards will render with empty panels for that visualization, as the visualization is not removed from the dashboards. Another example is you can have alerts that reference non-existing actions. One more example, is you can have a Dashboard drilldown that references a deleted dashboard.

Some ideas:

- Saved Object service could provide a *delete* method which checks if it is "safe to delete" that saved object, i.e. it has no other objects referencing it. Delete would fail if a saved object is referenced by one or more other saved objects.
- Saved Objects service could provide a first class way to retrieve all objects referencing your object. (Currently, it is also possible but is not part of Saved Objects service API).
- Not sure if we need it in Kibana, but there could be an option to execute *cascading delete* &mdash; it would delete the object you are trying to delete and all other objects referencing it, if any, recursively.

Related:

- [Alerting needs referential integrity](https://github.com/elastic/kibana/issues/45145)
